crates/wasi/src/preview2/host/tcp.rs

```diff
@@ -0,0 +1,571 @@
+            _ => panic!(),
+        };
+
+        // Report errors, resetting the state if needed.
+        match result {
+            Ok(()) => {}
+            Err(err) => {
+                *tcp_state = HostTcpState::Default;
+                return Err(err.into());
+            }
+        }
+
+        drop(tcp_state);
+
+        let input_clone = socket.clone_inner();
+        let output_clone = socket.clone_inner();
+
+        let input_stream = self.table_mut().push_input_stream(Box::new(input_clone))?;
+        let output_stream = self
+            .table_mut()
+            .push_output_stream(Box::new(output_clone))?;
+
+        Ok((input_stream, output_stream))
+    }
+
+    fn start_listen(&mut self, this: tcp::TcpSocket) -> Result<(), network::Error> {
+        let table = self.table();
+        let socket = table.get_tcp_socket(this)?;
+
+        let mut tcp_state = socket.inner.tcp_state.write().unwrap();
+        match &*tcp_state {
+            HostTcpState::Bound => {}
+            HostTcpState::ListenStarted => return Err(ErrorCode::AlreadyListening.into()),
+            HostTcpState::Connected => return Err(ErrorCode::AlreadyConnected.into()),
+            _ => return Err(ErrorCode::NotInProgress.into()),
+        }
+
+        socket.tcp_socket().listen(None)?;
+
+        *tcp_state = HostTcpState::ListenStarted;
+        socket.inner.sender.send(()).unwrap();
+
+        Ok(())
+    }
+
+    fn finish_listen(&mut self, this: tcp::TcpSocket) -> Result<(), network::Error> {
+        let table = self.table();
+        let socket = table.get_tcp_socket(this)?;
+
+        let mut tcp_state = socket.inner.tcp_state.write().unwrap();
+
+        match &mut *tcp_state {
+            HostTcpState::ListenStarted => {}
+            _ => return Err(ErrorCode::NotInProgress.into()),
+        }
+
+        let new_join = spawn_task_to_wait_for_connections(socket.clone_inner());
+        *tcp_state = HostTcpState::Listening(Pin::from(Box::new(new_join)));
+        drop(tcp_state);
+
+        Ok(())
+    }
+
+    fn accept(
+        &mut self,
+        this: tcp::TcpSocket,
+    ) -> Result<(tcp::TcpSocket, InputStream, OutputStream), network::Error> {
+        let table = self.table();
+        let socket = table.get_tcp_socket(this)?;
+
+        let mut tcp_state = socket.inner.tcp_state.write().unwrap();
+        match &mut *tcp_state {
+            HostTcpState::ListenReady(_) => {}
+            HostTcpState::Listening(join) => match maybe_unwrap_future(join) {
+                Some(joined) => joined.unwrap(),
+                None => return Err(ErrorCode::WouldBlock.into()),
+            },
+            HostTcpState::Connected => return Err(ErrorCode::AlreadyConnected.into()),
+            _ => return Err(ErrorCode::NotInProgress.into()),
+        }
+
+        let new_join = spawn_task_to_wait_for_connections(socket.clone_inner());
+        *tcp_state = HostTcpState::Listening(Pin::from(Box::new(new_join)));
```

Comment:
I guess we also set the state without making the pollable ready as well, so maybe there should be two methods for it? `set` and `set_and_wake`?

Reply:
The obvious `set` and `set_and_wake` would require dropping and re-acquiring the lock. We don't need to worry about threads here just yet, but when we do, we'll need to hold onto that lock. So I factored out some different utility functions here, in a way that should still help ensure that we get the sequence right.